File: Atcoder_#318/javascript/c_blue_spring.js
function main(input) {
    // 入力を配列に変換
    const lines = input.trim().split('\n');
    const [n, d, p] = lines[0].split(' ').map(Number);
    const f = lines[1].split(' ').map(Number);

    // 運賃を降順でソート
    f.sort((a, b) => b - a);

    // 最小コストを計算する
    let totalCost = 0;
    let i = 0;

    while (i < n) {
        // 期間内の運賃の合計を計算
        if (i + d <= n) {
            const periodCost = f.slice(i, i + d).reduce((sum, fare) => sum + fare, 0);

            // 周遊パスを使用した方が安い場合
            if (periodCost > p) {
                totalCost += p;
            } else {
                totalCost += periodCost;
            }
            i += d;
        } else {
            // 残りの日がd枚に満たない場合
            const remainingDays = n - i;
            const remainingCost = f.slice(i, i + remainingDays).reduce((sum, fare) => sum + fare, 0);

            // 残りの日数に対してもパスを使用するか判断
            if (remainingCost > p) {
                totalCost += p;
            } else {
                totalCost += remainingCost;
            }
            break;
        }
    }

    // 結果を出力
    console.log(totalCost);
}

// 標準入力を読み込み、main関数を呼び出し
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin', 'utf8');
main(input);

File: Atcoder_#318/ruby/c_blue_spring.rb
# 入力を受け取る
n, d, p = gets.split.map(&:to_i)
f = gets.split.map(&:to_i)

# 運賃を降順でソート
f.sort!.reverse!

# 最小コストを計算する
total_cost = 0
i = 0

while i < n
  # 期間内の運賃の合計を計算
  if i + d <= n
    period_cost = f[i, d].sum

    # 周遊パスを使用した方が安い場合
    if period_cost > p
      total_cost += p
    else
      total_cost += period_cost
    end
    i += d
  else
    # 残りの日がd枚に満たない場合
    remaining_days = n - i
    remaining_cost = f[i, remaining_days].sum

    # 残りの日数に対してもパスを使用するか判断
    if remaining_cost > p
      total_cost += p
    else
      total_cost += remaining_cost
    end
    break
  end
end

puts total_cost
